File: apps/windows/TeslaSync.App/feature-views/InfrastructureSection.Model.cs
using System.Text.Encodings.Web;
using System.Text.Json;
using TeslaSync.App.Core.Data.State;

namespace TeslaSync.App.FeatureViews.Infrastructure;

/// <summary>
/// The five backend tools the Infrastructure section exposes — the native union of the cards the web
/// <c>InfrastructureSection</c> renders (web/src/features/admin/components/devtools/InfrastructureSection.tsx):
/// four read-only <c>BackendTool</c>s (Db Stats / Migrations / Env Check / Runtime) and the bespoke MQTT
/// publish tool. Each maps to one generated dev-tools operation.
/// </summary>
public enum InfrastructureToolKind
{
    /// <summary><c>GET /dev-tools/db-stats</c> — database table/row statistics.</summary>
    DbStats,

    /// <summary><c>GET /dev-tools/migration-status</c> — applied/pending schema migrations.</summary>
    Migrations,

    /// <summary><c>POST /dev-tools/mqtt-test</c> — publish a test message to an MQTT topic.</summary>
    MqttTest,

    /// <summary><c>GET /dev-tools/env-check</c> — required environment/configuration check.</summary>
    EnvCheck,

    /// <summary><c>GET /dev-tools/runtime-info</c> — Go runtime / build / resource info.</summary>
    Runtime,
}

/// <summary>
/// The mutually-exclusive lifecycle state of a single tool card. Every value maps onto a visible surface
/// (none is hidden), mirroring the web tool's <c>useMutation</c> lifecycle: idle (no result yet → the
/// ResultPanel's "No result yet"), running (the button's <c>loading</c> spinner), succeeded (the success
/// badge + the JSON ResultPanel), and the failure split the web collapses into <c>data.error</c> —
/// <see cref="Failed"/> for a server/in-band error and <see cref="Offline"/> for a transport/offline fault.
/// </summary>
public enum InfrastructureToolStatus
{
    /// <summary>No run has been requested yet — the idle ResultPanel ("No result yet").</summary>
    Idle,

    /// <summary>A run is in flight — the button shows a progress ring; any prior result stays visible (stale).</summary>
    Running,

    /// <summary>The run returned data with no error — the success badge + the JSON ResultPanel.</summary>
    Succeeded,

    /// <summary>The run reported an error (HTTP non-success, in-band <c>error</c>, or server fault).</summary>
    Failed,

    /// <summary>The run failed because the API could not be reached — the offline-flavoured error.</summary>
    Offline,
}

/// <summary>
/// Static, vehicle-agnostic metadata for one tool card — the native mirror of a web
/// <c>BackendTool</c>/<c>MqttTestTool</c> invocation (its icon, accent, title/description i18n and the
/// dev-tools endpoint it runs). Resolved against the generated endpoint table
/// (<c>TeslaSync.Windows.Generated.Api.ApiEndpoints</c>) by <see cref="OperationId"/>; the accent and glyph
/// are token/font keys the view resolves so this stays UI-free and unit-testable.
/// </summary>
public sealed record InfrastructureToolDescriptor(
    InfrastructureToolKind Kind,
    string OperationId,
    bool RequiresInput,
    string Glyph,
    string AccentBrushKey,
    string TitleKey,
    string TitleFallback,
    string DescriptionKey,
    string DescriptionFallback)
{
    /// <summary>The five tools, in the exact order the web section lays them out.</summary>
    public static IReadOnlyList<InfrastructureToolDescriptor> Catalog { get; } = new[]
    {
        new InfrastructureToolDescriptor(
            InfrastructureToolKind.DbStats,
            "get_api_v1_dev_tools_db_stats",
            RequiresInput: false,
            Glyph: "\uE9F5", // StorageOptical
            AccentBrushKey: "TsColorInfoBrush",
            TitleKey: "featureView.infrastructure.dbStats.title",
            TitleFallback: "Db Stats",
            DescriptionKey: "featureView.infrastructure.dbStats.desc",
            DescriptionFallback: "Database size, table counts and row statistics."),
        new InfrastructureToolDescriptor(
            InfrastructureToolKind.Migrations,
            "get_api_v1_dev_tools_migration_status",
            RequiresInput: false,
            Glyph: "\uE9D5", // Flow
            AccentBrushKey: "TsColorSuccessBrush",
            TitleKey: "featureView.infrastructure.migrations.title",
            TitleFallback: "Migrations",
            DescriptionKey: "featureView.infrastructure.migrations.desc",
            DescriptionFallback: "Applied and pending database schema migrations."),
        new InfrastructureToolDescriptor(
            InfrastructureToolKind.MqttTest,
            "post_api_v1_dev_tools_mqtt_test",
            RequiresInput: true,
            Glyph: "\uE704", // Streaming
            AccentBrushKey: "TsColorWarningBrush",
            TitleKey: "featureView.infrastructure.mqtt.title",
            TitleFallback: "Mqtt",
            DescriptionKey: "featureView.infrastructure.mqtt.desc",
            DescriptionFallback: "Publish a test message to an MQTT topic."),
        new InfrastructureToolDescriptor(
            InfrastructureToolKind.EnvCheck,
            "get_api_v1_dev_tools_env_check",
            RequiresInput: false,
            Glyph: "\uEA18", // Shield
            AccentBrushKey: "TsChartPowerBrush",
            TitleKey: "featureView.infrastructure.envCheck.title",
            TitleFallback: "Env Check",
            DescriptionKey: "featureView.infrastructure.envCheck.desc",
            DescriptionFallback: "Required environment variables and configuration."),
        new InfrastructureToolDescriptor(
            InfrastructureToolKind.Runtime,
            "get_api_v1_dev_tools_runtime_info",
            RequiresInput: false,
            Glyph: "\uEC4A", // Processor
            AccentBrushKey: "TsColorWarningBrush",
            TitleKey: "featureView.infrastructure.runtime.title",
            TitleFallback: "Runtime",
            DescriptionKey: "featureView.infrastructure.runtime.desc",
            DescriptionFallback: "Go runtime, build and resource information."),
    };

    /// <summary>Find the descriptor for a <see cref="InfrastructureToolKind"/> (throws if unknown).</summary>
    public static InfrastructureToolDescriptor For(InfrastructureToolKind kind) =>
        Catalog.First(d => d.Kind == kind);
}

/// <summary>
/// The result of running one tool — the native analogue of the web <c>mutation.data</c> shape, where a
/// fulfilled <c>useMutation</c> yields either the parsed JSON body or an <c>{ error }</c> envelope. A
/// success carries the raw <see cref="Value"/>; a failure carries a localized-ready
/// <see cref="ErrorMessage"/> and the <see cref="ErrorKind"/> classification so the view can pick the
/// failed-vs-offline surface.
/// </summary>
public sealed record InfrastructureToolOutcome(
    bool Succeeded,
    JsonElement? Value,
    string? ErrorMessage,
    RepositoryErrorKind? ErrorKind)
{
    /// <summary>A successful run carrying the parsed response body.</summary>
    public static InfrastructureToolOutcome Success(JsonElement value) =>
        new(true, value.Clone(), null, null);

    /// <summary>A failed run carrying a privacy-safe message and its classification.</summary>
    public static InfrastructureToolOutcome Failure(string message, RepositoryErrorKind kind) =>
        new(false, null, message, kind);
}

/// <summary>
/// Pure, UI-free projections shared by the tool view-models and the headless tests — the JSON
/// pretty-printer (the web <c>JSON.stringify(data, null, 2)</c>), the in-band <c>data.error</c> reader
/// (the web's <c>mutation.data.error</c> truthiness check), the offline classification and the Narrator
/// name builder. Kept here so the per-state composition is verified without a XAML runtime.
/// </summary>
public static class InfrastructureToolProjection
{
    private static readonly JsonSerializerOptions IndentedJson = new()
    {
        WriteIndented = true,
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    /// <summary>Pretty-print a JSON value with two-space indentation (web <c>JSON.stringify(data, null, 2)</c>).</summary>
    public static string PrettyJson(JsonElement value) => JsonSerializer.Serialize(value, IndentedJson);

    /// <summary>
    /// Reproduce the web's <c>mutation.data.error</c> check: a response object whose <c>error</c> field is
    /// truthy is treated as a failure. <paramref name="message"/> is the error string when <c>error</c> is a
    /// (non-empty) string, otherwise null — matching the web's
    /// <c>typeof data.error === 'string' ? data.error : undefined</c>.
    /// </summary>
    public static bool TryReadInbandError(JsonElement value, out string? message)
    {
        message = null;
        if (value.ValueKind != JsonValueKind.Object ||
            !value.TryGetProperty("error", out var error))
        {
            return false;
        }

        switch (error.ValueKind)
        {
            case JsonValueKind.String:
                var text = error.GetString();
                if (string.IsNullOrEmpty(text))
                {
                    return false;
                }

                message = text;
                return true;
            case JsonValueKind.True:
                return true;
            case JsonValueKind.Number:
                return error.TryGetDouble(out var n) && n != 0;
            case JsonValueKind.Object:
            case JsonValueKind.Array:
                return true;
            default:
                return false;
        }
    }

    /// <summary>True when a failure was a transport/offline fault (the offline-flavoured surface).</summary>
    public static bool IsOffline(RepositoryErrorKind kind) =>
        kind is RepositoryErrorKind.Network or RepositoryErrorKind.Offline;

    /// <summary>The status a fulfilled outcome resolves to (success / in-band-error / offline / failed).</summary>
    public static InfrastructureToolStatus StatusFor(InfrastructureToolOutcome outcome)
    {
        ArgumentNullException.ThrowIfNull(outcome);
        if (outcome.Succeeded && outcome.Value is { } value)
        {
            return TryReadInbandError(value, out _) ? InfrastructureToolStatus.Failed : InfrastructureToolStatus.Succeeded;
        }

        return outcome.ErrorKind is { } kind && IsOffline(kind)
            ? InfrastructureToolStatus.Offline
            : InfrastructureToolStatus.Failed;
    }

    /// <summary>
    /// The Narrator description for a tool card: the tool title plus a state phrase so the live region
    /// announces "Db Stats: success", "Mqtt: failed", etc. Never includes raw response bodies (PII-safe).
    /// </summary>
    public static string AutomationName(string title, InfrastructureToolStatus status, string statePhrase) =>
        $"{title}: {statePhrase}";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/InfrastructureSection/InfrastructureSectionModel.kt
// Pure, framework-free model + projection for the Infrastructure dev-tools section — the native
// analogue of everything web/src/features/admin/components/devtools/InfrastructureSection.tsx (and its
// composed BackendTool / MqttTestTool / ResultPanel) derives before returning JSX. No Compose, no
// Android framework, no HTTP: every type here is unit-tested off device in the :android:testReleaseUnitTest
// gate, keeping the composable a thin render layer.
//
// The web surface is a grid of five on-demand dev-tools, each a `useMutation` (NOT a polling query): the
// operator presses Run, a single request fires, and its JSON result OR error is shown. db-stats /
// migration-status / mqtt-test / env-check / runtime-info map 1:1 to the five tools below.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/InfrastructureSection — the P3 prompt's allowed-files path) cannot form a
// valid Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the
// package intentionally diverges from the path — exactly as the sibling dashboard widgets do.
// `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.infrastructure

import io.teslasync.android.data.ErrorKind
import io.teslasync.android.data.errorKindOf
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull

/**
 * The accent color band of a tool card — the native analogue of the web `ICON_COLOR_MAP` keys passed to
 * each tool (`cyan` / `green` / `amber` / `purple`). Kept framework-free here; the render layer maps each
 * band onto a semantic [io.teslasync.android.components.ui.IconBoxTone] so light/dark/high-contrast all
 * resolve correctly (web neon hues have no 1:1 Material token, so they map to the nearest semantic tone).
 */
enum class InfraTone { Cyan, Green, Amber, Purple }

/**
 * The five dev-tools the web `InfrastructureSection` renders, in source order. Each carries the i18n key
 * for its title + description (the exact keys the web `t(...)` calls use), its accent [tone], and the
 * `/dev-tools/{endpoint}` request it issues. [needsInput] marks the MQTT tool, the only one with a body
 * (topic + message) — every other tool is a bodyless GET, mirroring the web `BackendTool` default.
 *
 * @property endpoint the path segment under `/dev-tools/` (web `apiFetch(endpoint, ...)`).
 * @property post whether the request is a POST (web MQTT tool) vs the GET default.
 * @property titleKey i18n key for the card title (web `t(title)`).
 * @property descKey i18n key for the card description (web `t(description)`).
 * @property tone the accent band (web `ICON_COLOR_MAP` color).
 * @property needsInput whether the tool collects a topic + message before running (web `MqttTestTool`).
 */
enum class InfraTool(
    val endpoint: String,
    val post: Boolean,
    val titleKey: String,
    val descKey: String,
    val tone: InfraTone,
    val needsInput: Boolean,
) {
    DbStats("db-stats", post = false, InfraKeys.DB_STATS, InfraKeys.DB_STATS_DESC, InfraTone.Cyan, needsInput = false),
    Migrations("migration-status", post = false, InfraKeys.MIGRATIONS, InfraKeys.MIGRATIONS_DESC, InfraTone.Green, needsInput = false),
    MqttTest("mqtt-test", post = true, InfraKeys.MQTT, InfraKeys.MQTT_DESC, InfraTone.Amber, needsInput = true),
    EnvCheck("env-check", post = false, InfraKeys.ENV_CHECK, InfraKeys.ENV_CHECK_DESC, InfraTone.Purple, needsInput = false),
    Runtime("runtime-info", post = false, InfraKeys.RUNTIME, InfraKeys.RUNTIME_DESC, InfraTone.Amber, needsInput = false),
}

/**
 * The i18n keys the web source passes to `t(...)`, verbatim. The natural-key style (`t('Db Stats')`) is
 * the web app's own convention here; some of these keys exist in the shared P1/S10 catalog (`Mqtt`,
 * `Migrations`, `Runtime`, `Topic`, `Message`, `Run`, `Success`, `Failed`, `Copy`, `Copied`) and the rest
 * fall back to the key text exactly as react-i18next does on the web. The render layer resolves each
 * through the Android resource facade, falling back to the key — reproducing i18next's behaviour 1:1
 * (see `resolveInfraText` in InfrastructureSection.kt) so the on-screen text matches the web verbatim.
 */
object InfraKeys {
    const val DB_STATS = "Db Stats"
    const val DB_STATS_DESC = "Db Stats Desc"
    const val MIGRATIONS = "Migrations"
    const val MIGRATIONS_DESC = "Migrations Desc"
    const val MQTT = "Mqtt"
    const val MQTT_DESC = "Mqtt Desc"
    const val ENV_CHECK = "Env Check"
    const val ENV_CHECK_DESC = "Env Check Desc"
    const val RUNTIME = "Runtime"
    const val RUNTIME_DESC = "Runtime Desc"

    const val TOPIC = "Topic"
    const val MESSAGE = "Message"
    const val SEND_TEST = "Send Test"
    const val RUN = "Run"
    const val SUCCESS = "Success"
    const val FAILED = "Failed"
    const val COPY = "Copy"
    const val COPIED = "Copied"
    const val NO_RESULT = "No result yet"
    const val OFFLINE = "common.offline"
    const val REQUEST_FAILED = "Request failed"
}

/** The lifecycle of a single tool's on-demand run — the native analogue of a web `useMutation` status. */
enum class RunPhase {
    /** Nothing has been run yet (web `!mutation.data && !mutation.isPending`). */
    Idle,

    /** A request is in flight (web `mutation.isPending`) — the Run button shows a spinner. */
    Running,

    /** The request returned a payload with no truthy `error` (web Success badge + result panel). */
    Succeeded,

    /** The request failed — transport error OR a 2xx body carrying a truthy `error` (web Failed badge). */
    Failed,
}

/**
 * Immutable result of one tool run — the projection of a single mutation outcome onto render-ready state.
 * Pure data (no Compose) so every branch is unit-tested directly.
 *
 * @property phase the current run phase.
 * @property result the decoded success payload to pretty-print (web `mutation.data`), else `null`.
 * @property errorDetail a verbatim error string when the backend returned `{error: "..."}` on a 2xx
 *   (web `typeof mutation.data.error === 'string' ? mutation.data.error : undefined`), else `null`; for a
 *   transport failure this is `null` and the render layer derives a localized message from [errorKind].
 * @property errorKind the classified transport failure (web's thrown-error branch), else `null`.
 */
data class ToolRun(
    val phase: RunPhase = RunPhase.Idle,
    val result: JsonElement? = null,
    val errorDetail: String? = null,
    val errorKind: ErrorKind? = null,
) {
    val isIdle: Boolean get() = phase == RunPhase.Idle
    val isRunning: Boolean get() = phase == RunPhase.Running
    val isSucceeded: Boolean get() = phase == RunPhase.Succeeded
    val isFailed: Boolean get() = phase == RunPhase.Failed

    /**
     * True when the most recent failure was a connectivity fault (no network / circuit open / timeout) —
     * the surface shows an "offline" chip and the Run button stays available to retry once back online.
     */
    val isOffline: Boolean
        get() =
            phase == RunPhase.Failed &&
                (errorKind == ErrorKind.Network || errorKind == ErrorKind.CircuitOpen || errorKind == ErrorKind.Timeout)

    companion object {
        val IDLE: ToolRun = ToolRun()
    }
}

/**
 * The whole section's state — one [ToolRun] per [InfraTool]. Each tool is independent (its own mutation),
 * exactly as each web tool owns its own `useMutation`.
 */
data class InfrastructureSectionState(
    val runs: Map<InfraTool, ToolRun> = InfraTool.entries.associateWith { ToolRun.IDLE },
) {
    /** The run state for [tool] (never null — defaults to [ToolRun.IDLE]). */
    fun runOf(tool: InfraTool): ToolRun = runs[tool] ?: ToolRun.IDLE

    /** Returns a copy with [tool]'s run replaced by [run]. */
    fun with(
        tool: InfraTool,
        run: ToolRun,
    ): InfrastructureSectionState = copy(runs = runs + (tool to run))

    companion object {
        fun initial(): InfrastructureSectionState = InfrastructureSectionState()
    }
}

/**
 * Canonical registry metadata for the Infrastructure surface — the native mirror of the web devtools
 * section. The diagnostics [SLUG] is emitted with the one-shot `view.opened` event (P1/S11).
 */
object InfrastructureSectionRegistration {
    /** Stable surface id. */
    const val ID: String = "devtools-infrastructure"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "InfrastructureSection"
}

/**
 * Pure projection from a mutation [Result] to render-ready [ToolRun] — the native port of the web
 * `mutation.data.error ? 'Failed' : 'Success'` / `data={...error ? undefined : data}` logic. Side
 * effect free so the gate unit-tests every branch without a device.
 */
object InfrastructureSectionProjection {
    private val PRETTY: Json =
        Json {
            prettyPrint = true
            prettyPrintIndent = "  "
        }

    /**
     * Projects a finished mutation [result] onto a [ToolRun]:
     *  - success whose body carries a truthy `error` → [RunPhase.Failed] with the error string (if a
     *    string) as [ToolRun.errorDetail], no result shown (web hides `data` when `error` is truthy);
     *  - success with no truthy `error` → [RunPhase.Succeeded] with the payload;
     *  - failure (transport) → [RunPhase.Failed] with the classified [ErrorKind] (web's catch branch,
     *    which surfaces `{error: err.message}`).
     */
    fun projectRun(result: Result<JsonElement>): ToolRun =
        result.fold(
            onSuccess = { json ->
                if (hasTruthyError(json)) {
                    ToolRun(phase = RunPhase.Failed, result = null, errorDetail = errorStringOrNull(json), errorKind = null)
                } else {
                    ToolRun(phase = RunPhase.Succeeded, result = json, errorDetail = null, errorKind = null)
                }
            },
            onFailure = { error ->
                ToolRun(phase = RunPhase.Failed, result = null, errorDetail = null, errorKind = errorKindOf(error))
            },
        )

    /**
     * Mirrors JavaScript truthiness of `data.error`: an object body whose `error` field is a non-empty
     * string, a non-zero number, boolean `true`, or any object/array is "failed"; `null`, `""`, `0`,
     * `false` (and a body that is not an object, or has no `error` field) are not.
     */
    fun hasTruthyError(json: JsonElement): Boolean {
        val error = (json as? JsonObject)?.get("error") ?: return false
        return when (error) {
            is JsonNull -> false
            is JsonPrimitive ->
                when {
                    error.isString -> error.content.isNotEmpty()
                    error.booleanOrNull != null -> error.booleanOrNull == true
                    error.doubleOrNull != null -> error.doubleOrNull != 0.0
                    else -> error.content.isNotEmpty()
                }
            is JsonObject -> true
            is JsonArray -> true
        }
    }

    /** The `error` value when it is a JSON string (web `typeof error === 'string'`), else `null`. */
    fun errorStringOrNull(json: JsonElement): String? {
        val error = (json as? JsonObject)?.get("error") as? JsonPrimitive ?: return null
        return if (error.isString && error.content.isNotEmpty()) error.content else null
    }

    /** Two-space pretty JSON, the native analogue of web `JSON.stringify(data, null, 2)`. */
    fun prettyJson(json: JsonElement): String = PRETTY.encodeToString(JsonElement.serializer(), json)
}
